csrc/optimization/consecutive_cast_pass.cpp

```diff
@@ -0,0 +1,99 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/utils.h>
+#include <optimization/consecutive_cast_pass.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+void castOptimizationPass(Fusion* fusion) {
+  auto is_cast_op = [](Expr* expr) {
+    if (expr->isA<UnaryOp>()) {
+      auto op = expr->as<UnaryOp>();
+      if (op->getUnaryOpType() == UnaryOpType::Cast) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  // NOTE: not the most efficient pass
+  std::unordered_map<Val*, Val*> replacement_map;
+  for (auto expr : fusion->exprs()) {
+    if (is_cast_op(expr)) {
+      bool mutated = false;
+      while (true) {
+        // in the loop, we just repetitively skip consecutive casts.
+        auto intermediate_cast = expr->input(0);
```

Comment:
I thought about factoring the usage of intermediate on the logic, but I'm not totally sure about `so skipping doesn't seem beneficial.`, so I didn't put that in.
i.e. for an arbitrary graph, even with an intermediate cast uses by some other nodes, the *other node* could have ended up in another segmented kernel?!

Admittedly, if it's a fusion output, it's likely and preferably going to be in the same fusion, so we might want to skip that.

I'll put that logic in.

Reply:
Because the intermediate needs to be produced, the difference before and after would just look like:

Before:
```
b <- cast(a)
c <- cast(b)
```

After:
```
b <- cast(a)
c <- cast(a)
```

So, in either way, both `b` and `c` are produced. I'm not sure why this transformation would be beneficial.

Reply:
I think this depends on how the fusion is segmented. But I agree that's only a trivial thing and doesn't really matter in a general case. I've already backed out this and now the cast folding stops when intermediate is either a real output or used by a non-cast operation (since cast operation could potentially be folded away as well).